Consider actual indentation of previous line for smart indent

Triton171 opened the pull request:
Essentially, to compute the indentation of a new line, we run an indent query both for the insertion position and for a non-empty line above. Then, we add the difference of these computed indents to the actual indentation of the line above. This makes indent queries a lot more resilient: Any error / inconsistency which applies to both lines will be canceled out, e.g.:
- Incorrect / Incomplete indent queries. In particular, incomplete indent queries are no longer worse than no indent queries with this change. This should also make it easier to contribute new languages since one can start by adding a few simple captures and then gradually add more.
- Incorrect parsing of the syntax tree by tree-sitter (e.g. if the code is incomplete)
- Different formatting styles where only one can correctly be handled by indent queries. This is especially interesting for languages like Markdown and LaTex where indent queries weren't really feasible until now due to the variation in formatting.

The main thing still missing from this PR is some way to disable relative indent queries. While this shouldn't really be necessary when using Helix, relative indent computation makes debugging indent queries more difficult. Should this be an option? Or is there a better way to do it?
I also haven't tested my changes a lot, so I appreciate it if anyone wants to try it out. I'll also use it myself for some time before marking the PR ready for merging.

The only significant downside of this change is that we now compute 2 (or in rare cases 3) indent queries when inserting a new line. The time needed for this only becomes noticable when inserting a new line with many cursors at once. If this is an issue, we could disable smart indent when inserting many (e.g. > 1000) newlines at once. For now, I think it is fine though (I haven't seen any complaints about the current indent queries being slow with many cursors either).

Fixes #6235
Fixes #7206
Fixes (partially) #6939 and #1725

clo4 commented:
After I manually closed #8124, I realised it shouldn't have been because plain text is still an issue - this PR would also have closed that, given that it works great with plain text. Fantastic contribution, looking forward to when this is merged! It feels great to use.

pascalkuthe commented on helix-core/src/indent.rs:
```diff
@@ -849,24 +936,76 @@ pub fn indent_for_newline(
+                // a constant number of (non-empty) lines.
+                const MAX_ATTEMPTS: usize = 2;
+                let mut num_attempts = 0;
+                for line_idx in (0..=line_before).rev() {
```
Instead of iterating line indecies you can use `text.lines_at(line_before + 1).reversed()` to iterate lines (it's faster/more ideomatic)

pascalkuthe replied:
You could also handle the max attempt by adding `take(MAX_ATTEMPTS)`

Triton171 replied:
Just iterating lines doesn't work here because `treesitter_indent_for_pos` needs a line index. Using `take(MAX_ATTEMPTS)` would work but it would slightly change the logic: Currently, blank lines are not counted towards the number of attempts, since we don't do any indent computation on them. If we did count blank lines, this would cause us to fallback to the `tree-sitter` heuristic much more often.

pascalkuthe replied:
Fair enough, it's a bit slower to iterate lines this way (about 5x) but it doesn't matter too much here I supposed.

Triton171 commented on book/src/configuration.md:
```diff
@@ -65,6 +65,7 @@ Its settings will be merged with the configuration directory `config.toml` and t
 | `workspace-lsp-roots` | Directories relative to the workspace root that are treated as LSP roots. Should only be set in `.helix/config.toml` | `[]` |
 | `default-line-ending` | The line ending to use for new documents. Can be `native`, `lf`, `crlf`, `ff`, `cr` or `nel`. `native` uses the platform's native line ending (`crlf` on Windows, otherwise `lf`). | `native` |
 | `insert-final-newline` | Whether to automatically insert a trailing line-ending on write if missing | `true` |
+| `indent-heuristic` | How the indentation for a newly inserted line is computed: `simple` just copies the indentation level from the previous line, `tree-sitter` computes the indentation based on the syntax tree and `hybrid` combines both approaches | `hybrid`
```
Is there a way to set a language setting for all languages at once? I think that the main use case of this option will be for debugging changes to the indent queries / heuristic, where changing the config for each relevant language might be tedious.

In case a language doesn't have queries, the `simple` heuristic is always used as a fallback (`hybrid` falls back to `tree-sitter` and `tree-sitter` falls back to `simple`). Buggy / low quality queries also shouldn't be a bug issue anymore. An empty indent query file with the `hybrid` heuristic should be essentially equivalent to using the `simple` heuristic and each added pattern will generally only improve the experience.
The only issue are incorrect patterns that e.g. often apply `@indent` where it isn't needed. Such patterns should never be added in the first place though, so I'm not sure if we need a mechanism to work around them.

pascalkuthe replied:
Ah that's even better than I missed that there was such a fallback on my first read. Now it's clearer that there is documentation for it 👍 

It is possible to create language settings with a global fallback but it does indeed seem like it's not needed here

Triton171 commented:
Thanks for the reminder that this PR still exists :D. I rebased it onto the latest master and added 2 alignment queries for C, mainly since they aid the new hybrid indent heuristic in choosing a suitable baseline. Since I've daily-driven this for the past few months at work (doing C++ development), I'm reasonably confident that the code shouldn't have any major bugs.

Some other languages might also still have missing alignment queries which could cause issues with the new indent heuristic. I think these should be rare, so we can just fix them whenever they are reported (there's not really a way to find them other than using Helix to edit code in the language for a significant amount of time).

archseer reviewed (approved):
As always, thanks for improving our indentation capabilities! 🎉
